manta-pay/src/crypto/poseidon/encryption.rs

```diff
@@ -0,0 +1,304 @@
+    PartialEq(bound = "S::Field: PartialEq")
+)]
+pub struct CiphertextBlock<S, COM = ()>(Box<[S::Field]>)
+where
+    S: Specification<COM>;
+
+impl<S, COM> Write<Permutation<S, COM>, COM> for CiphertextBlock<S, COM>
+where
+    S: Specification<COM>,
+    S::Field: Clone + BlockElement<COM>,
+{
+    type Output = PlaintextBlock<S, COM>;
+
+    #[inline]
+    fn write(&self, state: &mut State<S, COM>, compiler: &mut COM) -> Self::Output {
+        for (i, elem) in state.iter_mut().skip(1).enumerate() {
+            *elem = self.0[i].sub(elem, compiler);
+        }
+        PlaintextBlock(state.iter().skip(1).cloned().collect())
+    }
+}
+
+/// Authentication Tag
+#[cfg_attr(
+    feature = "serde",
+    derive(Deserialize, Serialize),
+    serde(
+        bound(
+            deserialize = "S::Field: Deserialize<'de>",
+            serialize = "S::Field: Serialize"
+        ),
+        crate = "manta_util::serde",
+        deny_unknown_fields
+    )
+)]
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = "S::Field: Clone"),
+    Debug(bound = "S::Field: Debug"),
+    Eq(bound = "S::Field: Eq"),
+    Hash(bound = "S::Field: Hash"),
+    PartialEq(bound = "S::Field: PartialEq")
+)]
+pub struct Tag<S, COM = ()>(S::Field)
+where
+    S: Specification<COM>;
+
+impl<S, COM> Read<Permutation<S, COM>, COM> for Tag<S, COM>
+where
+    S: Specification<COM>,
+    S::Field: Clone,
+{
+    #[inline]
+    fn read(state: &State<S, COM>, compiler: &mut COM) -> Self {
+        let _ = compiler;
+        Self(state.0[1].clone())
+    }
+}
+
+/// Encryption Configuration
+#[cfg_attr(
+    feature = "serde",
+    derive(Deserialize, Serialize),
+    serde(
+        bound(
+            deserialize = "S::Field: Deserialize<'de>",
+            serialize = "S::Field: Serialize"
+        ),
+        crate = "manta_util::serde",
+        deny_unknown_fields
+    )
+)]
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = "S::Field: Clone"),
+    Debug(bound = "S::Field: Debug"),
+    Eq(bound = "S::Field: Eq"),
+    Hash(bound = "S::Field: Hash"),
+    PartialEq(bound = "S::Field: PartialEq")
+)]
+pub struct Encryption<S, COM = ()>
+where
+    S: Specification<COM>,
+{
+    /// Initial State
+    pub initial_state: State<S, COM>,
+}
+
+impl<S, COM> Types<Permutation<S, COM>, COM> for Encryption<S, COM>
+where
+    S: Specification<COM>,
+    S::Field: Clone + BlockElement<COM>,
+{
+    type Key = Vec<S::Field>;
+    type Header = Vec<S::Field>;
+    type SetupBlock = SetupBlock<S, COM>;
+    type PlaintextBlock = PlaintextBlock<S, COM>;
+    type CiphertextBlock = CiphertextBlock<S, COM>;
+    type Tag = Tag<S, COM>;
+}
+
+impl<S, COM> Setup<Permutation<S, COM>, COM> for Encryption<S, COM>
+where
+    S: Specification<COM>,
+    S::Field: Clone + Default + BlockElement<COM>,
+{
+    #[inline]
+    fn initialize(&self, compiler: &mut COM) -> State<S, COM> {
+        let _ = compiler;
+        self.initial_state.clone()
```

Comment:
This can be set to whatever the consumer of the encryption API wants it to be (this is a setting on the encryption struct). It can be all zeros or can be something else.